File: 2020/day08.cpp
#include <cstdio>
#include <set>
#include <string>
#include <vector>

using namespace std;
  
pair<bool, int> runProg(const vector<pair<string, int>>& prog) {
  set<int> prevExecuted;

  int pc = 0, acc = 0;
  while(pc >= 0 && pc < prog.size() && !prevExecuted.count(pc)) {
    auto inst = prog[pc];
    prevExecuted.insert(pc);

    if(inst.first == "acc") acc += inst.second;
    else if(inst.first == "jmp") pc += inst.second - 1;
    pc++;
  }
  return {pc == prog.size(), acc};
}

pair<bool, int> fixProg(vector<pair<string, int>>& prog) {
  for(auto& inst : prog) {
    if(inst.first != "acc") {
      inst.first = inst.first == "nop" ? "jmp" : "nop";
      auto res = runProg(prog);
      if(res.first) return {true, res.second};
      inst.first = inst.first == "nop" ? "jmp" : "nop";
    }
  }
  return {false, 0};
}

int main() {
  vector<pair<string, int>> prog;
  char op[5]; int arg;
  while(scanf("%s %d", op, &arg) > 0) {
    prog.emplace_back(op, arg);
  }
  int res1 = runProg(prog).second;
  int res2 = fixProg(prog).second;
  printf("%d %d\n", res1, res2);
  return 0;
}

File: 2020/day08.hs
import Control.Arrow
import Data.Sequence (Seq)
import qualified Data.Sequence as Seq
import qualified Data.Set as Set
import Text.Regex

runProg :: Seq (String, Int) -> (Bool, Int)
runProg = runProg' Set.empty 0 0
  where
    runProg' visited pc acc prog
      | pc >= Seq.length prog = (True, acc)
      | pc `Set.member` visited = (False, acc)
      | otherwise =
        let next pc' acc' = runProg' (Set.insert pc visited) pc' acc' prog
         in case prog `Seq.index` pc of
              ("acc", a) -> next (pc + 1) (acc + a)
              ("jmp", a) -> next (pc + a) acc
              ("nop", _) -> next (pc + 1) acc
              (opcode, _) -> error $ "Illegal opcode: " <> opcode

solve1 :: Seq (String, Int) -> Int
solve1 = snd . runProg

solve2 :: Seq (String, Int) -> Int
solve2 prog = head $ map snd $ filter fst $ map (runProg . fixProg) [0 .. Seq.length prog - 1]
  where
    fixProg idx = Seq.adjust' fixInst idx prog
    fixInst ("nop", a) = ("jmp", a)
    fixInst ("jmp", a) = ("nop", a)
    fixInst inst = inst

parseInput :: String -> Seq (String, Int)
parseInput = Seq.fromList . map readInst . lines
  where
    readInst str =
      case matchRegex (mkRegex "([a-z]+) \\+?([-0-9]+)") str of
        Just [opcode, arg] -> (opcode, read arg)
        _ -> error $ "Invalid instruction: " <> str

main :: IO ()
main = interact $ show . (solve1 &&& solve2) . parseInput
